README.md
# moment-opening-times

[![GitHub Release](https://img.shields.io/github/release/nhsuk/moment-opening-times.svg)](https://github.com/nhsuk/moment-opening-times/releases/latest/)
[![npm version](https://badge.fury.io/js/moment-opening-times.svg)](https://badge.fury.io/js/moment-opening-times)
[![Greenkeeper badge](https://badges.greenkeeper.io/nhsuk/moment-opening-times.svg)](https://greenkeeper.io/)
[![Build Status](https://github.com/nhsuk/moment-opening-times/workflows/CI/badge.svg)](https://github.com/nhsuk/moment-opening-times/actions)
[![Coverage Status](https://coveralls.io/repos/github/nhsuk/moment-opening-times/badge.svg?branch=master)](https://coveralls.io/github/nhsuk/moment-opening-times?branch=master)

> Given a set of opening times (with optional alternative opening times) return a status indicating openness.

## Installation

To install the package run:

* npm - `npm install moment-opening-times --save`
* yarn - `yarn add moment-opening-times`

This package relies upon several other packages. These are the
[Peer Dependencies](https://nodejs.org/en/blog/npm/peer-dependencies/).
To get a list of the peerDependencies run the following command:

```
npm info moment-opening-times peerDependencies
```

All peerDependencies need to be installed.

## Usage

The class is instantiated with 3 arguments. The third argument is optional and
represents a set of opening times classified as alterations. Once instantiated
the function `getStatus` is passed the current time and an optional options object.
Currently the only valid option is a boolean field - `next`. When
`next` set to a truthy value the returned object will contain 2 additional
fields, `nextClosed` and `nextOpen`. Both fields are a
[moment with a time zone](https://momentjs.com/timezone/docs/#/using-timezones/parsing-in-zone/)
indicating the next time the place is closed and open, respectively.

### Example usage

Given the open hours of a place is represented by the following set of opening times:
```js
const openingTimes =
{
  sunday:    [ { opens: '09:00', closes: '17:30' } ],
  monday:    [ { opens: '08:00', closes: '11:00' }, { opens: '13:00', closes: '15:00' }, { opens: '17:00', closes: '19:00' } ],
  tuesday:   [ { opens: '09:00', closes: '17:30' } ],
  wednesday: [ { opens: '09:00', closes: '12:00' }, { opens: '14:00', closes: '18:00' } ],
  thursday:  [ { opens: '09:00', closes: '17:00' }, { opens: '18:00', closes: '21:00' } ],
  friday:    [ { opens: '09:00', closes: '17:30' } ],
  saturday:  [ { opens: '09:00', closes: '17:30' } ],
}
```

In order to find out if the place is open now in the `Europe/London` time zone
the following code can be used:
```js
const moment = require('moment');
const OpeningTimes = require('moment-opening-times');

const now = moment();
const openingTimesMoment = new OpeningTimes(openingTimes, 'Europe/London');

const status = openingTimesMoment.getStatus(now);
```

If we wanted to know whether the place was open at 23:59:59 on 31/12/2020 in
`Indian/Christmas` the code would be:
```js
const moment = require('moment');
const OpeningTimes = require('moment-opening-times');

const dateToCheck = moment('2020-12-31 23:59:59');
const openingTimesMoment = new OpeningTimes(openingTimes, 'Indian/Christmas');

const status = openingTimesMoment.getStatus(dateToCheck);
```

For more detailed use cases, check out the [test suite](test/).


CHANGELOG.md
2.10.0 / 2019-09-30
===================
- Update npm dependencies
- Upgrade to latest `eslint` versions and fix errors
- Use latest Husky hooks pattern
- Replace `istanbul` with `nyc` for code coverage
- Use ADRs to record architectural decisions
- Replace Travis CI with GitHub Actions
- Remove `coveralls`
- Remove `moment` (it is included in `moment-timezone`)

2.9.0 / 2018-04-13
==================
- Update npm dependencies
- Remove snyk

2.8.0 / 2018-03-15
==================
- Update to `eslint-config-nhsuk@0.13.0`

2.7.0 / 2018-03-13
==================
- Update npm dependencies

2.6.0 / 2018-01-02
==================
- Update npm dependencies

2.5.1 / 2017-11-01
==================
- Remove prepublish script

2.5.0 / 2017-11-01
==================
- Update npm dependencies

2.4.0 / 2017-09-05
==================
- Add CHANGELOG

2.3.0 / 2017-08-17
==================
- Removal of unnecessary usages of timeZone
- A simplification of moment creation
- Removal of unnecessary object creation
- Removal of unnecessary calculations

2.2.0 / 2017-08-14
==================
- Remove past alterations from being included in checks for future alterations

2.1.0 / 2017-07-26
==================
- Correct version number
- Publish to npm
- Improvements to README
- Use noun-verb pattern for npm scripts

< 2.1.0 / 2017-07-19
==================
- Initial work on the module


test/timings.js
const orgLimit = 1000;

const OpeningTimes = require('../OpeningTimes');
const orgs = require('./pharmacy-list').slice(0, orgLimit);
const Moment = require('moment');

const timeZone = 'Europe/London';

/* Notes:
 * These tests are for comparative timings and were used during performance optimisations
 * They have no assertions since they were for timing purposes only
 * Left in the code base as the basis for something which can be used during
 * CI to detect step changes in code execution time.
 */

describe('timings', function test() {
  this.timeout(6000);
  describe(`getStatus for ${orgs.length} orgs`, () => {
    it('disregarding alterations', () => {
      orgs.forEach((o) => {
        if (o.openingTimes) {
          const openingTimes =
            new OpeningTimes(o.openingTimes.general, timeZone);
          openingTimes.getStatus(
            new Moment('2016-11-05T11:00:00').tz(timeZone), { next: true });
        }
      });
    });

    it('including alterations', () => {
      orgs.forEach((o) => {
        if (o.openingTimes) {
          const openingTimes =
            new OpeningTimes(o.openingTimes.general, timeZone, o.openingTimes.alterations);
          openingTimes.getStatus(
            new Moment('2016-11-05T11:00:00').tz(timeZone), { next: true });
        }
      });
    });
  });
});


test/lib/utils.js
const alterationDateKeyFormat = 'YYYY-MM-DD';

// eslint-disable-next-line no-underscore-dangle
function _alterations(moment, daysToChange) {
  const dateString = moment.clone().add(daysToChange, 'day').format(alterationDateKeyFormat);

  const returnValue = {};
  returnValue[dateString] = [];
  return returnValue;
}

function alterationsPast(moment) {
  return _alterations(moment, -1);
}

function alterationsPresent(moment) {
  return _alterations(moment, 0);
}

function alterationsFuture(moment) {
  return _alterations(moment, 1);
}

function alterationsPresentAndFuture(moment) {
  return Object.assign(alterationsPresent(moment), alterationsFuture(moment));
}

function alterationsPastPresentAndFuture(moment) {
  return Object.assign(
    alterationsPast(moment),
    alterationsPresent(moment),
    alterationsFuture(moment)
  );
}

module.exports = {
  alterationsFuture,
  alterationsPast,
  alterationsPastPresentAndFuture,
  alterationsPresent,
  alterationsPresentAndFuture,
};


lib/parsers/nhs-choices-syndication/parser.js
const xml2js = require('xml2js');
const jsonQuery = require('json-query');
const stripPrefix = require('xml2js/lib/processors').stripPrefix;
const assert = require('assert');
const Verror = require('verror');

const parseOpeningTimesFromSyndicationXml = (openingTimesType, xml) => {
  assert(openingTimesType, 'parameter \'openingTimesType\' undefined/empty');
  assert(xml, 'parameter \'xml\' undefined/empty');
  assert.equal(typeof (openingTimesType),
    'string', 'parameter \'openingTimesType\' must be a string');
  assert.equal(typeof (xml),
    'string', 'parameter \'xml\' must be a string');
  const openingTimes = {};
  const options = {
    tagNameProcessors: [stripPrefix],
  };
  const xmlParser = new xml2js.Parser(options);
  xmlParser.parseString(xml, (err, result) => {
    if (err) {
      throw new Verror(err, 'Unable to parse opening times XML');
    }

    let openingTimesForType = null;
    try {
      if (result.feed.entry[0].content[0].overview[0].openingTimes) {
        const allOpeningTimes =
          result.feed.entry[0].content[0].overview[0].openingTimes[0].timesSessionTypes[0];

        openingTimesForType = jsonQuery('timesSessionType[*:isType]', {
          data: allOpeningTimes,
          locals: {
            isType: (item) => item.$.sessionType === openingTimesType,
          },
        }).value[0];

        openingTimesForType.daysOfWeek[0].dayOfWeek.forEach((item) => {
          const dayName = item.dayName[0].toLowerCase();
          openingTimes[dayName] = [];
          item.timesSessions[0].timesSession.forEach((t) => {
            if (t.fromTime) {
              // session details are a time range
              openingTimes[dayName].push({ opens: t.fromTime[0], closes: t.toTime[0] });
            }
          });
        });
      }
    } catch (e) {
      throw new Verror(e, `Unable to get '${openingTimesType}' opening times from xml`);
    }
  });
  return openingTimes;
};

module.exports = parseOpeningTimesFromSyndicationXml;


OpeningTimes.js
const assert = require('assert');
const util = require('util');
const Moment = require('moment-timezone');
const removePastAlterations = require('./src/lib/removePastAlterations');

const weekdays = Moment.weekdays().map((d) => d.toLowerCase());

class OpeningTimes {
  constructor(openingTimes, timeZone, alterations) {
    assert(openingTimes, 'parameter \'openingTimes\' undefined/empty');
    const parameterWeek = Object.keys(openingTimes).sort();
    assert.deepEqual(
      parameterWeek, weekdays.sort(),
      `parameter 'openingTimes' should have all days of the week (${parameterWeek})`
    );
    assert(
      weekdays.every((d) => Array.isArray(openingTimes[d])),
      'parameter \'openingTimes\' should define opening times for each day.'
      + ` (${util.inspect(openingTimes)})`
    );

    assert(timeZone, 'parameter \'timeZone\' undefined/empty');
    assert(Moment.tz.zone(timeZone), 'parameter \'timeZone\' not a valid timezone');
    this._openingTimes = openingTimes;
    this._timeZone = timeZone;
    this._alterations = alterations;
  }

  /* Private methods - you could use them but they  are not part of the API
     and therefore any changes to them will not trigger a major version number */

  _getDayName(moment) {
    return moment.format('dddd').toLowerCase();
  }

  _getTime(moment, hour, minute) {
    const returnDate = moment.clone().tz(this._timeZone);
    returnDate.startOf('hour').hour(hour).minute(minute);

    return returnDate;
  }

  _getTimeFromString(timeString) {
    const timeSplit = timeString.split(':');
    return {
      hours: parseInt(timeSplit[0], 10),
      minutes: parseInt(timeSplit[1], 10),
    };
  }

  _createDateTime(moment, timeString) {
    const time = this._getTimeFromString(timeString);
    return this._getTime(moment, time.hours, time.minutes);
  }

  _isClosedAllDay(daysOpeningTimes) {
    return (daysOpeningTimes.length === 0);
  }

  _formatTime(timeString, formatString = 'h:mm a') {
    if (timeString === '00:00' || timeString === '23:59') {
      return 'midnight';
    }
    const aDate = Moment();
    const time = this._getTimeFromString(timeString);
    return this._getTime(aDate, time.hours, time.minutes).format(formatString);
  }

  _getOpeningTimesForDate(moment) {
    const alterations = removePastAlterations(this._alterations, moment);
    if (alterations) {
      // TODO: decide what to do if there is only >1 match
      const momentDate = moment.format('YYYY-MM-DD');
      const todaysAlteration = Object.keys(alterations).find((a) => a === momentDate);

      return todaysAlteration
        ? alterations[todaysAlteration]
        : this._openingTimes[this._getDayName(moment)];
    }
    return this._openingTimes[this._getDayName(moment)];
  }

  _getOpeningTimesSessionForMoment(moment, daysLookAhead) {
    for (let day = daysLookAhead - 1; day >= -1; day -= 1) {
      const aMoment = moment.clone().add(day, 'day');
      const openingTimes = this._getOpeningTimesForDate(aMoment);

      for (let j = 0; j < openingTimes.length; j += 1) {
        const t = openingTimes[j];
        const from = this._createDateTime(aMoment, t.opens);
        const to = this._createDateTime(aMoment, t.closes);

        if (to < from) {
          to.add(1, 'day');
        }

        if (moment.isBetween(from, to, null, '[)')) {
          return { from, to };
        }
      }
    }
    return undefined;
  }

  _getOpenSessions(moment, days) {
    // Get sessions for week (including yesterday for the cases where opening hours span midnight)
    return days.map((d) => {
      const aDate = moment.clone().add(d, 'day');
      const openingTimes = this._getOpeningTimesForDate(aDate);
      return openingTimes
        .map((t) => {
          const from = this._createDateTime(aDate, t.opens);
          const to = this._createDateTime(aDate, t.closes);

          if (to < from) {
            to.add(1, 'day');
          }

          return { from, to };
        });
    });
  }

  _getDateInSessionFinder(moment) {
    return (session) => (moment.isBetween(session.from, session.to, null, '[)'));
  }

  _getDateBeforeSessionFinder(moment) {
    return (session) => (moment.isBefore(session.from));
  }

  _findMomentInSessions(moment, sessions) {
    return sessions.some((day) => (day.some(this._getDateInSessionFinder(moment))));
  }

  _findNextOpeningTime(moment, sessions) {
    const nextDay = sessions
      .filter((day) => (!this._isClosedAllDay(day)))
      .find((day) => (day.some(this._getDateBeforeSessionFinder(moment))));

    if (nextDay) {
      return nextDay.find(this._getDateBeforeSessionFinder(moment)).from;
    }

    return undefined;
  }

  _nextOpen(moment) {
    const allSessions = this._getOpenSessions(moment, [0, 1, 2, 3, 4, 5, 6]);
    if (this._findMomentInSessions(moment, allSessions)) {
      return moment;
    }
    return this._findNextOpeningTime(moment, allSessions);
  }

  /* Public API */

  getStatus(moment, options = {}) {
    const returnValue = { moment };

    const session = this._getOpeningTimesSessionForMoment(moment, 1);
    returnValue.isOpen = (session !== undefined);
    if (options.next) {
      if (returnValue.isOpen) {
        returnValue.nextClosed = session.to;
        returnValue.nextOpen = moment;
      } else {
        returnValue.nextClosed = moment;
        returnValue.nextOpen = this._nextOpen(moment);
      }
    }

    return returnValue;
  }

  getFormattedOpeningTimes(formatString) {
    const openingTimes = {};

    Moment.weekdays().forEach((d) => {
      const day = d.toLowerCase();
      openingTimes[day] = this._openingTimes[day].map((t) => (
        {
          closes: this._formatTime(t.closes, formatString),
          opens: this._formatTime(t.opens, formatString),
        }));
    });

    return openingTimes;
  }
}

module.exports = OpeningTimes;
